pdns/dnsdist-lua-actions.cc

```diff
@@ -489,6 +489,65 @@ class LuaFFIAction: public DNSAction
   func_t d_func;
 };
 
+class LuaFFIPerThreadAction: public DNSAction
+{
+public:
+  typedef std::function<int(dnsdist_ffi_dnsquestion_t* dq)> func_t;
+
+  LuaFFIPerThreadAction(const std::string& code): d_functionCode(code), d_functionID(s_functionsCounter++)
+  {
+  }
+
+  DNSAction::Action operator()(DNSQuestion* dq, std::string* ruleresult) const override
+  {
+    try {
+      auto& state = t_perThreadStates[d_functionID];
+      if (!state.d_initialized) {
+        setupLuaFFIPerThreadContext(state.d_luaContext);
+        state.d_func = state.d_luaContext.executeCode<func_t>(d_functionCode);
```

Comment:
if the `executeCode` fails, what's get assigned to `state.d_func` ?

Reply:
The default value is an empty function, which should yield `std::bad_function_call ` on later invocation. That should not happen since the state would not be marked as initialized so we would retry again and again. That's not very nice but I'm not sure we can do better, except perhaps by marking the state as initialized before calling `executeCode` and checking later that the function is not empty, so we only try to execute it once. Any thoughts?

Reply:
I'm torn. Multiple error messages is indeed ugly, I like a single error message better  But: having faulty Lua is a bad thing, and might event warrant a fatal error...

Reply:
I don't want a fatal error to happen at runtime, so I guess what I will do is log a warning only once.